### DISCUSSION TITLE
Allow for inheritance of devcontainer spec

### DISCUSSION THREAD

**Discussion by benjaminbauer** [devcontainers/community#9] [Ideas]
### Use case
Working on a project that utilizes devcontainer with a team. The preferences wrt to tooling differs among team members (zsh vs fish, vim vs neovim, etc.).
### Current solution
make the devcontainer a union of all the requirments. I.e. install all tools, that any developer in the team favors.
### Challenges

1. Container will become bloated
2. potential conflicts between requirements

#### `include` property
add an optional property `include` to thee devcontainer spec. The value is a valid devcontainer file. When defined, the container described in that file will be the base for the focal definition. I.e. I can add my own tools on top of that definition.

Base file: checked in the project repo, contains the project-wide standards (runtime, linting tools, etc.)
extending file: developer preferences, e.g. certain vscode plugins, editors, shells, etc.

##### Challenges
conflicts between base and extending definition (e.g. different image)
#### devcontainer extension file
To avoid the complexity of inheritance as above a special extension file format can be defined. This would be a subset of the devcontainer spec. The properties would only be able to append entries to the base file

example:
```
{
      "parent": "./devcontainer.json"
      "customizations": {
		"vscode": {
			"extensions": [
				"ms-azuretools.vscode-docker"
			]
		}
	}
}
```
This would add `ms-azuretools.vscode-docker` to the `parent`

**Comment by Chuxel** [comment]
Great suggestion! There's a proposal for this at https://github.com/devcontainers/spec/issues/22. See what you think!

**Reply by benjaminbauer** [reply]
Exactly the problem that I am raising here. Thanks for the hint!

**Reply by Chuxel** [reply]
Great! There's even a PR out for it right now in the reference implementation: https://github.com/devcontainers/cli/pull/311

**Comment by benjaminbauer** [comment]
When working with devcontainers in VS Code (which is probably true for most users), the described use case can be partially solved with:
1. https://code.visualstudio.com/docs/devcontainers/containers#_always-installed-extensions
2. and the analogous setting for features `dev.containers.defaultFeatures`

What I would like in the same spirit is: https://github.com/microsoft/vscode-dev-containers/issues/1221

Aside of https://github.com/devcontainers/cli/pull/311, this seems to support this use case too: https://github.com/microsoft/vscode-remote-release/issues/3279

**Comment by nilroy** [comment]
Any update on this? This is a very needed feature. Other solutions like `flox` can handle this.

**Comment by Falco20019** [comment]
We also really wait for this. Would make differentiating between local dev and CI a lot easier. CI could just skip all `customizations` and skip some mounts / commands only needed for local dev that would otherwise highly increase the time building takes.
